cudax/include/cuda/experimental/__utility/basic_any/conversions.cuh

```diff
@@ -0,0 +1,149 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __CUDAX_DETAIL_BASIC_ANY_CONVERSIONS_H
+#define __CUDAX_DETAIL_BASIC_ANY_CONVERSIONS_H
+
+#include <cuda/std/detail/__config>
+
+#include "cuda/std/__cccl/attributes.h"
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__type_traits/conditional.h>
+#include <cuda/std/__type_traits/is_convertible.h>
+#include <cuda/std/__type_traits/remove_reference.h>
+
+#include <cuda/experimental/__utility/basic_any/basic_any_fwd.cuh>
+#include <cuda/experimental/__utility/basic_any/interfaces.cuh>
+
+_CCCL_PUSH_MACROS
+#undef interface
+
+namespace cuda::experimental
+{
+//!
+//! conversions
+//!
+//! Can one basic_any type convert to another? Implicitly? Explicitly?
+//! Statically? Dynamically? We answer these questions by mapping two
+//! cvref qualified basic_any types to archetype types, and then using
+//! the built-in language rules to determine if the conversion is valid.
+//!
+struct __immovable_archetype
+{
+  __immovable_archetype()                        = default;
+  __immovable_archetype(__immovable_archetype&&) = delete;
+
+  template <class _Value>
+  _CUDAX_HOST_API __immovable_archetype(_Value) noexcept;
+  template <class _Value>
+  _CUDAX_HOST_API __immovable_archetype(_Value*) = delete;
+};
+
+struct __movable_archetype : __immovable_archetype
+{
+  __movable_archetype() = default;
+  _CUDAX_HOST_API __movable_archetype(__movable_archetype&&) noexcept;
+};
```

Comment:
Nitpick: I personally prefer to be more explicit by deleting the SMF that I do not want exposed